src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+        auto it = index.lower_bound(EntryPeer{peer, false, UINT256_ZERO});
+        while (it != index.end() && it->m_peer == peer) {
+            // Check what to continue with after this iteration. Note that 'it' may change position, and
+            // std::next(it) may be deleted in the process, so this needs to be decided beforehand.
+            auto it_next = (std::next(it) == index.end() || std::next(it)->m_peer != peer) ?
+                index.end() : std::next(it);
+            // If the entry isn't already COMPLETED, first make it COMPLETED (which will mark other CANDIDATEs as
+            // CANDIDATE_BEST, or delete all of a txhash's entries if no non-COMPLETED ones are left).
+            if (MakeCompleted(m_index.project<ByTxHash>(it))) {
+                // Then actually delete the entry (unless it was already deleted by MakeCompleted).
+                Erase<ByPeer>(it);
+            }
+            it = it_next;
+        }
+    }
+
+    void ForgetTxHash(const uint256& txhash)
+    {
+        auto it = m_index.get<ByTxHash>().lower_bound(EntryTxHash{txhash, State::CANDIDATE_DELAYED, 0});
+        while (it != m_index.get<ByTxHash>().end() && it->m_txhash == txhash) {
+            it = Erase<ByTxHash>(it);
+        }
+    }
+
+    void ReceivedInv(uint64_t peer, const GenTxid& gtxid, bool preferred,
+        std::chrono::microseconds reqtime)
+    {
+        // Bail out if we already have a CANDIDATE_BEST entry for this (txhash, peer) combination. The case where
+        // there is a non-CANDIDATE_BEST entry already will be caught by the uniqueness property of the ByPeer index
+        // automatically.
+        if (m_index.get<ByPeer>().count(EntryPeer{peer, true, gtxid.GetHash()})) return;
+
+        // Find last entry for this txhash, and extract per-txhash information from it.
+        Iter<ByTxHash> it_last = m_index.get<ByTxHash>().end();
+        // First find the first entry past this txhash.
+        it_last = m_index.get<ByTxHash>().lower_bound(EntryTxHash{gtxid.GetHash(), State::TOO_LARGE, 0});
+        if (it_last != m_index.get<ByTxHash>().begin() && std::prev(it_last)->m_txhash == gtxid.GetHash()) {
+            it_last--;
+        } else {
+            // No entry for this txhash exists yet.
+            it_last = m_index.get<ByTxHash>().end();
+        }
+
+        // Try creating the entry with CANDIDATE_DELAYED state (which will fail due to the uniqueness
+        // of the ByPeer index if a non-CANDIDATE_BEST entry already exists with the same txhash and peer).
+        // Bail out in that case.
+        auto ret = m_index.get<ByPeer>().emplace(gtxid, peer, preferred, reqtime, m_sequence);
+        if (!ret.second) return;
+
+        // Update accounting metadata.
+        ++m_peerinfo[peer].m_total;
+        ++m_sequence;
+    }
+
+    //! Find the GenTxids to request now from peer.
+    std::vector<GenTxid> GetRequestable(uint64_t peer, std::chrono::microseconds now)
+    {
```

Comment:
This function is now on the hot path and is called for every peer on every SendMessages() loop. Have you done any profiling to see how much time we'll spend in here on a normally loaded system? I think all of the lookups in here are O(1) in the size of the index, but is the constant factor important?

Reply:
`m_index.get<ByX>.begin()` should be `O(log_2(n))`, `++it` should be `O(1)`, `.emplace()` should be `O(log_2(n))` with about 3x constant factor overhead, `.modify()` should the same except maybe 6x constant factor overhead. `n` is limited to about 600k, so `log_2(n)` is below about 20. With max peers and max announcements, the entire index could be something like 100MB I think, more if you're removing the bounds for some peers via the relay permission.

I think worst case complexity is that in SetTimePoint you end up adjusting the state of every announcement , then you find the first BEST for the given peer, which turns out to be every tx for that peer, then iterate through the remaining BEST for that peer adding them to a vector, then sort the vector, and construct the result. I think that's about:

    600000 * (20 + 2*20) + 20 + 5000 + 5000*12 + 5000 ~= 36M ops

but assuming time doesn't go backwards, then amortized across multiple calls to `GetRequstable`, each announcement only causes 7 transitions over its lifetime (delayed, ready, best(+demotion of other), requested, completed(+promotion of other), and only 4 of those occur via `GetRequestable` so amortized cost per tx is something like 4*60+20+1+1*12+1 = 274 ops, so max average cost of calling `GetRequestable` should be about 1.4M ops.

On a normally loaded system, you should be getting no more than 35-70 txs from each call to GetRequestable, and I think SetTimePoint should likewise be dealing with (far) fewer than 7000 txs in each call, so load should be about 100 to 1000 times less, I think.

Reply:
I'll do some benchmarks.

Reply:
In a TxRequestTracker with 125 peers, and 5000 (unique) txids each, in a stable state (where `GetRequestable` returns nothing), each call takes around 100 ns.

Reply:
Another benchmark: I tweaked the unit test to run 12000 times as many test cases, but with the sanity checks turned off, and then timed the actual execution of the queued actions, and while gathering statistics about the number of calls, and the Size() of the data structure:
Runtime: 22.2403 s
* Size(): 622234 +- 175404
* ReceivedInvs calls: 6494712
* RequestedTx calls: 678552
* ReceivedResponse calls: 2663691
* DisconnectedPeer calls: 3491745
* ForgetTxHash calls: 145404
* GetRequestable calls: 35819152

So the size is relatively close to 625000 (the limit at 5000 tx/peer, 125 peers). The scenarios themselves are not representative for this, as they have far fewer transactions per peer. However, I believe there is reason to assume that the number of tx/peer doesn't matter (except for the peak latency of DisconnectedPeer, but averaged out, it's still amortized).

So this means that the average time per announcement (summed over all calls affecting that announcement) in these scenarios (which may be nonrepresentative) is around 3.4 us/announcement.

Reply:
These seem like good numbers.

I was going to suggest moving`SetTimePoint()` out of `GetRequestable()` and only calling it once per message handler thread loop (e.g. in a global tasks function[[[ like https://github.com/bitcoin/bitcoin/pull/19364/commits/0ea9abf9b4c3694dede390b759df01c1ce0d3166), since it doesn't need to be called in the context of an individual peer, and it doesn't need to be called frequently. However, it doesn't seem like there's a good reason to do that from a performance standpoint.

Reply:
Moving `SetTimePoint()` out of `GetRequestable()` might have better worst-case behaviour in the event of arbitrarily non-monotonic clocks In particular, currently you could have, say, two peers, peer=1 with no announcements, and peer=2 with N announcements all with a reqtime = t. If you call `GetRequestable(peer=1, now=t+1)` then you'll get an empty vector as the result, but update all N announcements to CANDIDATE_BEST, but if time goes backwards, and you then call `GetRequestable(peer=2, now=t-1)` you'll update all the announcements back to CANDIDATE_DELAYED, and all get an empty vector, and make no progress. If you did `SetTimePoint()` first, then called `GetRequestable()` for each peer without updating the timepoint, you'd either have `now=t-1` and not do any work, or you'd have `now=t+1` and would make forward progress. I think the constraint would be "call SetTimePoint, then for each peer call GetRequestable, and for each resulting tx, call RequestedTx for it" with the result being something along the lines of "amortized time complexity is `O(log(max_concurrent_announcements))`".

FWIW, I think something along those lines is worth considering as a post-merge TODO -- perhaps as part of doing a fuzz tester for how net_processing uses txrequest or something, but don't think there is a need to hold things up for it.

Reply:
Hmm, that's interesting. Before the `RequestedTx` change I'd be hesitant about such a change, as it'd make specifying the conditions under which it can be called even more cumbersome. There isn't much concern about that anymore, though.

@ajtowns Right, or more generally: if your clock is arbitrarily jittery, increasing the frequency of SetTimePoint calls will result in up to proportionally more work.

There may be some impact how strong the "first non-delayed announcement" effect is, though (which was the reason for dropping the "first" marker rule): if SetTimePoint is only called once per loop, the request will go to any of the peers that announced a txhash in one loop, rather than the actual first one. This may need some investigation.

So indeed, I'd keep this for a follow-up.

Reply:
Marking as resolved.